Index: crun/src/libcrun/container.c
===================================================================
--- crun.orig/src/libcrun/container.c
+++ crun/src/libcrun/container.c
@@ -2652,6 +2652,18 @@ libcrun_container_run_internal (libcrun_
 
   /* The container is waiting that we write back.  In this phase we can launch the
      prestart hooks.  */
+  /* Do this without being the subreaper for anything a hook spawns.
+   * XXX: is there a better way? (as if the container exits here, we won't catch it)
+  */
+  if (! detach || context->notify_socket)
+    {
+      ret = prctl (PR_SET_CHILD_SUBREAPER, 0, 0, 0, 0);
+      if (UNLIKELY (ret < 0))
+        {
+          ret = crun_make_error (err, errno, "set child subreaper");
+          goto fail;
+        }
+    }
   if (def->hooks && def->hooks->prestart_len)
     {
       libcrun_debug ("Running 'prestart' hooks");
@@ -2668,6 +2680,15 @@ libcrun_container_run_internal (libcrun_
       if (UNLIKELY (ret != 0))
         goto fail;
     }
+  if (! detach || context->notify_socket)
+    {
+      ret = prctl (PR_SET_CHILD_SUBREAPER, 1, 0, 0, 0);
+      if (UNLIKELY (ret < 0))
+        {
+          ret = crun_make_error (err, errno, "set child subreaper");
+          goto fail;
+        }
+    }
 
   if (seccomp_fd >= 0)
     {
